### PR TITLE
fix: fix get-func-path command

### PR DESCRIPTION
**Pull request by meifans**
E2E TEST: https://github.com/OfficeDev/TeamsFx/actions/runs/2359510927

### CODE REVIEW

**Comment by kuojianlu** [comment]
image
Update the comment as well to keep consistent.

File: packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts [hunk 1]
```diff
@@ -248,7 +248,7 @@ function startBackend(programmingLanguage: string): Record<string, unknown> {
     options: {
       cwd: "${workspaceFolder}/api",
       env: {
-        PATH: "${env:PATH}${command:fx-extension.get-func-path}",
+        PATH: "${command:fx-extension.get-func-path}${env:PATH}",
```
Comment (kuojianlu) [inline packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts:251]: How do we handle the existing projects?

**Comment by kuojianlu** [comment]
Update the CLI part as well: https://github.com/OfficeDev/TeamsFx/blob/main/packages/cli/src/cmds/preview/preview.ts#L1113.

**Comment by XiaofuHuang** [comment]
Will it affect the projects created in previous versions before we use prerequisite check task?

**Comment by XiaofuHuang** [comment]
Will it affect the projects created before we enable prerequisite check?
image

File: packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts (hunk 1 above)
Comment (qinezh) [reply packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts:251]: Another solution 
For node v16:
* If global func v3 installed, stop F5 with error message to ask user upgrade it to v4 manually. 
* If portable func v3 installed, upgrade func to v4 automatically and continue F5.

File: packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts [hunk 2]
```diff
@@ -341,7 +341,7 @@ function startFuncHostedBot(
     options: {
       cwd: "${workspaceFolder}/bot",
       env: {
-        PATH: "${env:PATH}${command:fx-extension.get-func-path}",
+        PATH: "${command:fx-extension.get-func-path}${env:PATH}",
```
Comment (a1exwang) [inline packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts:344]: Also need to change `fx-solution/debug/util/tasks.ts` to handle existing project add capability?

File: packages/vscode-extension/src/handlers.ts [hunk 3]
```diff
@@ -1422,7 +1422,7 @@ export async function getFuncPathHandler(): Promise<string> {
     const vscodeDepsChecker = new VSCodeDepsChecker(vscodeLogger, vscodeTelemetry);
     const funcStatus = await vscodeDepsChecker.getDepsStatus(DepsType.FuncCoreTools);
     if (funcStatus?.details?.binFolders !== undefined) {
-      return `${path.delimiter}${funcStatus.details.binFolders.join(path.delimiter)}`;
+      return `${funcStatus.details.binFolders.join(path.delimiter)}${path.delimiter}`;
```
Comment (swatDong) [inline packages/vscode-extension/src/handlers.ts:1425]: Add `${path.delimiter}` to both front and end.

Comment (meifans) [reply packages/vscode-extension/src/handlers.ts]: Added.

File: packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts (hunk 2 above)
Comment (meifans) [reply packages/fx-core/src/plugins/solution/fx-solution/debug/util/tasksNext.ts:344]: `fx-solution/debug/util/tasks.ts` is not used and it does use `get-func-path`. 

image

File: packages/cli/src/cmds/preview/preview.ts [hunk 4]
```diff
@@ -1281,6 +1281,7 @@ export default class Preview extends YargsCommand {
     let result = true;
     let summaryMsg = doctorResult.NodeSuccess;
     let helpLink = undefined;
+    let isNode12Installed = false;
```
Comment (swatDong) [inline packages/cli/src/cmds/preview/preview.ts:1284]: What's `isNode12Installed` for?

File: packages/fx-core/src/common/deps-checker/internal/funcToolChecker.ts [hunk 5]
```diff
@@ -82,6 +82,9 @@ export class FuncToolChecker implements DepsChecker {
       this._telemetry.sendEvent(DepsCheckerEvent.funcAlreadyInstalled, {
         "global-func-version": `${await this.queryGlobalFuncVersion()}`,
       });
+      if (!isPortableFuncInstalled) {
+        await this.cleanup();
```
Comment (swatDong) [inline packages/fx-core/src/common/deps-checker/internal/funcToolChecker.ts:86]: Will `cleanup` throw error if portable folder not exist?

File: packages/vscode-extension/src/handlers.ts [hunk 6]
```diff
@@ -1419,10 +1419,16 @@ export async function backendExtensionsInstallHandler(): Promise<string | undefi
  */
 export async function getFuncPathHandler(): Promise<string> {
   try {
+    if (!vscodeHelper.isFuncCoreToolsEnabled()) {
+      return "";
```
Comment (swatDong) [inline packages/vscode-extension/src/handlers.ts:1423]: Will `path.delimiter` be more compatible?

File: packages/cli/src/cmds/preview/preview.ts (hunk 4 above)
Comment (meifans) [reply packages/cli/src/cmds/preview/preview.ts:1284]: fixed.

File: packages/fx-core/src/common/deps-checker/internal/funcToolChecker.ts (hunk 5 above)
Comment (meifans) [reply packages/fx-core/src/common/deps-checker/internal/funcToolChecker.ts:86]: Not.
image

File: packages/vscode-extension/src/handlers.ts (hunk 6 above)
Comment (meifans) [reply packages/vscode-extension/src/handlers.ts]: Added

**Comment by meifans** [comment]
> Will it affect the projects created before we enable prerequisite check? image

We not change

> image Update the comment as well to keep consistent.

Changed.

**Comment by meifans** [comment]
> Update the CLI part as well: https://github.com/OfficeDev/TeamsFx/blob/main/packages/cli/src/cmds/preview/preview.ts#L1113.

Added.

File: packages/cli/src/cmds/preview/preview.ts [hunk 7]
```diff
@@ -1110,7 +1110,9 @@ export default class Preview extends YargsCommand {
     let funcEnv = undefined;
     if (func.details.binFolders !== undefined) {
       funcEnv = {
-        PATH: `${process.env.PATH}${path.delimiter}${func.details.binFolders.join(path.delimiter)}`,
+        PATH: `${path.delimiter}${func.details.binFolders.join(path.delimiter)}${process.env.PATH}${
```
Comment (swatDong) [inline packages/cli/src/cmds/preview/preview.ts:1113]: Missing delimiter between func-binFolders and env-PATH.

Comment (swatDong) [reply packages/cli/src/cmds/preview/preview.ts:1113]: Here can remove the redundant heading and ending delimiters.

**Comment by meifans** [comment]
> Will it affect the projects created before we enable prerequisite check? image

The way of `fx-extension.validate-dependencies` use `command` that always try to use portable func first and global func second. So I think it have no effect on previous versions.

File: packages/cli/src/cmds/preview/preview.ts (hunk 7 above)
Comment (meifans) [reply packages/cli/src/cmds/preview/preview.ts]: Fixed.

**Review by swatDong — APPROVED** [review]

**Review by kimizhu — APPROVED** [review]

**Review by zhenjiao-ms — APPROVED** [review]

**Pull request merged**